### PR TITLE
get unfrozen string before modifying

### PR DESCRIPTION
**Pull request by pderouen**
I was unable to create a case to duplicate the error. But, there is a case where the object passed to <code>clean_flag</code> is frozen (immutable). The string created by <code>to_s</code> will also be frozen in that case. The <code>.dup</code> method returns a mutable copy of the string.

### CODE REVIEW

**Comment by info-rchitect** [comment]
@pderouen Do you want to add a unit test to check for this corner case?  Otherwise LGTM.

File: program/flow_control_flag_bug.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+  log "Mixed-case manual flags"
+  test :test1, on_fail: { set_flag: :$My_Mixed_Flag }, continue: true, number: 51420
+  test :test2, if_flag: "$My_Mixed_Flag", number: 51430
+  unless_flag "$My_Mixed_Flag" do
+    test :test3, number: 51440
+  end
+  
+  log "Mixed-case manual flags - induce frozen string error"
+  test :test4, on_fail: { set_flag: :$My_Mixed_Flag }, continue: true, number: 51450
```
Comment (pderouen) [inline program/flow_control_flag_bug.rb:19]: Using the same flag name with a second set_flag is what leads to the frozen string bug.

**Comment by pderouen** [comment]
@info-rchitect I'm not sure how to get at the <code>clean_flag</code> method from a spec test. But, I was finally able to duplicate the behavior and added an example for it.

**Review by info-rchitect — APPROVED** [review]
LGTM

**Comment by redxeth** [comment]
@pderouen for this bug, what do you expect the proper output should look like? 

for UF today the `flow_control_flag_bug.rb` flow generates  :

~~~text           
Opcode    Parameter      Fail                Sense   Condition     Name 
Test      test1          My_Mixed_Flag_0
defaults                 My_Mixed_Flag
flag-true My_Mixed_Flag                              flag-true     My_Mixed_Flag_0
Test      test2                                      flag-true     My_Mixed_Flag
Test      test3                              not     flag-true     My_Mixed_Flag
Test      test4          My_Mixed_Flag_1
flag-true My_Mixed_Flag                              flag-true     My_Mixed_Flag_1
Test      test5                                      flag-true     My_Mixed_Flag
Test      test6                              not     flag-true     My_Mixed_Flag
~~~

It appears that the on_fail/set_flag is properly incrementing the flag name, but there are some unnecessary lines (defaults, flag-true lines) as well as the wrong flag used for test2/3/5/6 .  **Is that your expectation also?**

So it seems more than just the 'frozen strings' (`My_Mixed_Flag` not updating to include `_0` and `_1` but it seems other stuff is extraneous.  

Also, looking at it now I'm not entirely sure that `if_ran`/`unless_ran` also works?

@priyavadan @ginty @chrisnappi @info-rchitect

**Comment by pderouen** [comment]
@redxeth that uflex output produces the functionality that was specified in the test. This PR was just to resolve a hard crash. The flag string was intentionally frozen. When the _0 and _1 flag strings were being created by duplicating the frozen string, which produced frozen strings that couldn't have the _0 etc appended. Fixing that was the full scope of this PR. None of the underlying functionality was touched.

You should consider starting a new [discussion](https://github.com/Origen-SDK/origen_testers/discussions/landing).
